File: MAP/Seminar6_7/src/sem4/repository/InMemoryRepository.java
package sem4.repository;

import sem4.domain.HasID;
import sem4.domain.validator.Validator;

import java.util.HashMap;
import java.util.Map;

public class InMemoryRepository<ID, E extends HasID<ID>> implements CrudRepository<ID,E> {

    private Validator<E> validator;
    Map<ID,E> entities;

    public InMemoryRepository(Validator<E> validator) {
        this.validator = validator;
        entities=new HashMap<ID,E>();
    }

    @Override
    public E findOne(ID id){
        if (id==null)
            throw new IllegalArgumentException("id must be not null");
        return entities.get(id);
    }

    @Override
    public Iterable<E> findAll() {
        return entities.values();
    }

    @Override
    public E save(E entity) {
        if (entity==null)
            throw new IllegalArgumentException("entity must be not null");
        validator.validate(entity);
        if(entities.get(entity.getID()) != null) {
            return entity;
        }
        entities.put(entity.getID(),entity);
        return null;
    }

    @Override
    public E delete(ID id) {
        return null;
    }

    @Override
    public E update(E entity) {

        if(entity == null)
            throw new IllegalArgumentException("entity must be not null!");
        validator.validate(entity);
        if(entities.get(entity.getID()) != null) {
            entities.put(entity.getID(),entity);
            return null;
        }
        return entity;

    }

}

File: MAP/Seminar10/Sem10_MAP_223/Sem10_MAP_223/repository/InMemoryRepository.cs
using Sem10_MAP_223.model;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Sem10_MAP_223.repository
{
    class InMemoryRepository<ID, E> : IRepository<ID, E> where E:IHasID<ID>
    {

        private IDictionary<ID, E> entities = new Dictionary<ID, E>();
        private IValidator<E> validator;


        public InMemoryRepository(IValidator<E> vali)
        {
            this.validator = vali;
        }
        public E Delete(ID id)
        {
            throw new NotImplementedException();
        }

        public IEnumerable<E> FindAll()
        {
            return entities.Values.ToList();
        }

        public E FindOne(ID id)
        {
            throw new NotImplementedException();
        }

        public E Save(E entity)
        {
            if (entity == null)
                throw new ArgumentNullException("entity must not be null");
            validator.Validate(entity);
            if (entities.ContainsKey(entity.ID))
                return entity;
            entities[entity.ID] = entity;
            return default(E);
        }

        public E Update(E entity)
        {
            throw new NotImplementedException();
        }
    }
}
